File: weekly-contest/weekly-contest-131/1-remove-outermost-parentheses/delta.js
/**
 * +1/-1
 */
var removeOuterParentheses = function (S) {
  let delta = 0
  let res = ''
  for (let i = 0; i < S.length; ++i) {
    if (S[i] === '(' && delta !== 0 || S[i] === ')' && delta !== 1) {
      res += S[i]
    }
    if (S[i] === '(') ++delta
    else if (S[i] === ')') --delta
  }
  return res
};

[
  '(()())(())',
  '(()())(())(()(()))',
  '()()',
].forEach(s => {
  console.log(removeOuterParentheses(s))
})

File: problems/495-teemo-attacking/delta.cpp
/**
 * 离散化：`O(NlogN)`
 */

#include <vector>
#include <map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findPoisonedDuration(vector<int>& timeSeries, int duration) {
        map<int, int> delta;
        for (int start : timeSeries) {
            ++delta[start];
            --delta[start + duration];
        }

        int res = 0;
        int d = 0;
        int prev = -1;
        for (auto p : delta) {
            int curr = p.first;
            if (d > 0) res += curr - prev; // 流血
            prev = curr;
            d += p.second;
        }
        return res;
    }
};